### DISCUSSION TITLE
Undo search selection and couple other questions

### DISCUSSION THREAD

**Discussion by underlow** [marlonrichert/zsh-autocomplete#470] [Q&A] [answered]
Thank you for plugin, finally I've found a good one. Still I have several questions. 
1. I have live history search on and it's main use-case for me, but sometimes I need autocomplete. now when I type something and want to switch from history to autocomplete I have to press ctrl-r and cmd-space. Is there a way to bind it to one key? For example if pressing cmd-space in live history mode would switch to autocomplete mode would be great. In other words would be good to have hotkey to switch from like history to autocomplete and back 
2.  If you accidentally press arrow up/down some line from history is selected and there is no way to undo it. And it's pretty inconvenient. Maybe I missing something and there is a way?
3. Would be great have some commands (or strings) so that history switches to autocomplete automatically, for example if I type 'cd<space>' I would like to use directories autocomplete instead of history search. Auto switching would save some time for me. Is it possible?

**Comment by marlonrichert** [comment, marked as answer]
1. You don't need to press <kbd>Ctrl</kbd><kbd>Space</kbd>. Just <kbd>Ctrl</kbd><kbd>R</kbd> is enough to toggle between completion and history search. However, depending on your system, there might be a slight delay sometimes. Just be patient. 🙂 
2. Depending on your terminal, press <kbd>Ctrl</kbd><kbd>-</kbd> or <kbd>Ctrl</kbd><kbd>/</kbd> to undo. It's the default keybinding for undo in Zsh (unless you're using Vim mode).
3. It's not, but that is something that could be implemented. I created #471 for it.

**Reply by underlow** [reply]
1. Doesn't work for me. waited for 10+ seconds. 
2. It does work, thank you. But after Ctrl- or Ctrl/ it switches to autocomplete mode. Is there a hotkey to switch it back to live history?
3. 👍

**Reply by underlow** [reply]
Check item 1 several times - sometimes it switches sometimes not. Anything I can do to fix it?

**Reply by underlow** [reply]
I've got it. It does switch to completion mode, but I want next step. 
I'll change my question. 
- I type for example 'git' it shows history entries with git substring. 
<img width="178" alt="image" src="https://user-images.githubusercontent.com/15636461/186366512-e2663932-a75a-4cde-94fb-fbf3ecff7c20.png">
- After pressing Ctrl-R it switches to zsh-autosuggest (probably) with gray suggestion  
 
<img width="174" alt="image" src="https://user-images.githubusercontent.com/15636461/186366576-e0149283-40e2-494f-97a1-5f30131ecea5.png">
- And after Cmd-Space it switches to another mode (I don't know it's name) with command parameters suggestion 
<img width="628" alt="image" src="https://user-images.githubusercontent.com/15636461/186366828-d509fb06-7f6b-4ebf-aef7-dc46a00475e6.png">

Is it possible to bind some hotkey to go straight to 3 step?

**Reply by marlonrichert** [reply]
Step 2 doesn't happen to me. When I press <kbd>Ctrl</kbd><kbd>R</kbd>, it goes straight to step 3. Please [open a bug report](https://github.com/marlonrichert/zsh-autocomplete/issues/new?assignees=&labels=bug&template=bug-report.md&title=) for this and make sure you fill it out very precisely.

> But after Ctrl- or Ctrl/ it switches to autocomplete mode. Is there a hotkey to switch it back to live history?

Do you have `zstyle ':autocomplete:*' default-context history-incremental-search-backward` in your `.zshrc` file? Then Zsh Autocomplete will start in history search mode by default. Once in history mode, press <kbd>Ctrl</kbd><kbd>R</kbd> again to go back to completion mode.

**Reply by underlow** [reply]
I do have `zstyle ':autocomplete:*' default-context history-incremental-search-backward` in .zshrc

I'll clean up my .zshrc then test it once again and fill bug report if needed. Thank you

**Reply by underlow** [reply]
I'm trying to create a bug request but after executing script from Steps to reproduce section there is no live history mode. What I'm doing wrong?

**Reply by marlonrichert** [reply]
You need to either press <kbd>Ctrl</kbd><kbd>R</kbd> to activate it or run `zstyle ':autocomplete:*' default-context history-incremental-search-backward` to enable it by default.

**Reply by underlow** [reply]
Created https://github.com/marlonrichert/zsh-autocomplete/issues/473
if I can do anything to help with investigating, please I'm glad to help
